### ISSUE TITLE
Thrifty and default requiredness

### ISSUE THREAD

**Issue by luqasn** [microsoft/thrifty#420]
Is there any reason why Thrifty does not support the Thrift concept of default requiredness? 
Unless I am missing something (which might very well be the case), all it would take is making the current struct data class constructor private and adding a new public one where the fields with default requiredness are non-nullable. Also, throw in the builder upon building for fields with default requiredness.
That way, we have data classes with nullable fields for default requiredness (satisfying the “optional on read” concept from thrift), but you can only construct them by providing a value (“required on write”). The adapters would be unchanged and continue to treat default requiredness as optional.

**Comment by benjamin-bader** [comment]
That's an interesting take on default requiredness!

As for why Thrifty didn't do default requiredness, here's [what I wrote about it in 2015](https://github.com/microsoft/thrifty/blob/master/thrifty-schema/src/main/kotlin/com/microsoft/thrifty/schema/Requiredness.kt#L40).  Without a doubt it's the longest comment I've ever written.  Five years later it's less clear than I'd hoped; some details escape my memory.  The TL;DR of it is that not even the official Apache implementations seem to agree on what the default semantics are.  The de-facto "more equal than the others" implementation seems to be c++, and that one definitely has the behavior you described.  Others are more... pathological.

Basically this part of Thrift is un/under-specified, and since I didn't see a way to do it cleanly in a way that preserved the simplicity of the "null if absent, not null if set" model, I chose to leave it out.

It sounds like you've spotted a viable path forward, except for at this point there's five years of usage that assumes "default requiredness == optional".  I'm very reluctant to break people's generated code without good cause.  To that end, you're the first person to ever ask about this.  What brought it to mind for you?

**Comment by benjamin-bader** [comment]
And here is perhaps a clearer explanation of "y u no default": https://github.com/microsoft/thrifty/issues/34

**Comment by luqasn** [comment]
What brought this to mind is that I was adding server support to Thrifty (see over at [my fork](https://github.com/microsoft/thrifty/compare/master...luqasn:server-support if you are interested, but I wanna have a separate discussion about that as I know that is going to be controversial) and in that vein I was looking at making Thrifty as close to spec as possible.

That said:
* for structs used in responses, I don't see any value for default requiredness - required would have the same effect
* for structs used in requests, my understanding of default requiredness is that it is a way to say "hey, this is gonna be required soon, please start setting it from now on". Which sounds like a good idea at first, but you kinda ask yourself: if the server can handle the parameter not being there - as it must with default requiredness - e.g. with a default or something, why not keep it optional forever?
* we use thrift as a normal RPC protocol, but also (ab)use it in a different way: we use thrift structs without the methods part for event serialization and if a service consumes its own events and you want to introduce a new field to an event, making it have default requiredness would make sure the compiler tells you about the places where you should be setting it while you still maintain the ability to read older events with the field not present.

Thinking all of this out loud there really seems to be no big benefit. The way I described it it would be easy to add, so I would say "why not?" rather than "why?", but breaking people's builds is a cost that is too big for the little potential gain. But as you are preparing v3.0.0, couldn't it be included there as a breaking change? I could prepare a PR.

Anyways, this is not hugely important, so I would understand and respect if you don't wanna do it.

**Comment by benjamin-bader** [comment]
> adding server support to Thrifty
:open_mouth: no kidding!

>  couldn't it be included there as a breaking change?
Yes and no.  This kind of change would be far more disruptive to existing codebases than what we're contemplating for v3 so far.  Not to say that it _couldn't_ happen, but I'd want the benefit to justify the additional work we'd make for current users.  I don't (yet) see enough value.

Most of the breakage we're doing thus far is more on the dependency side, like requiring Java 8 compatibility, or defaulting the compiler to produce Kotlin code, or making the Kotlin multiplatform.  The runtime is also changing around a but.  What we _haven't_ done much of (any of?  I'd have to go check) is in breaking the generated code.  From experience with Outlook, changing the semantics there is _really_ painful.  It can be done, but again, there needs to be a good reason.

> adding server support to Thrifty

Coming back to that little morsel - if you'd rather wait until later to talk about it then that's fine, but I can't help but wonder... why Thrifty?  The Apache runtime is so much more capable and battle-tested that it just never occurred to me that anyone might want this.

**Comment by luqasn** [comment]
> Yes and no. This kind of change would be far more disruptive to existing codebases than what we're contemplating for v3 so far. Not to say that it _couldn't_ happen, but I'd want the benefit to justify the additional work we'd make for current users. I don't (yet) see enough value.

I agree. Could still see it being added as a flag, though. Maybe later™️ :)

> > adding server support to Thrifty
> 
> Coming back to that little morsel - if you'd rather wait until later to talk about it then that's fine, but I can't help but wonder... why Thrifty? The Apache runtime is so much more capable and battle-tested that it just never occurred to me that anyone might want this.

I am a bit surprised by that question (no offense), especially since you are now going Kotlin first on Thrifty. Aren't the reasons similar? Native Kotlin service/structs give you compile time null safety, safer & more convenient union handling and coroutine support.
Plus without the `isXXXSet` shenanigans you can properly use something like mapstruct to map your boundary to to your domain objects.

I see how "Server Support" does not really fit with the "Thrift for Android that saves you methods" tagline, but I think Thrifty already evolved beyond that. I don't think we (as in my team at my employer's) are not the only ones that wish this existed. Sidenote: we are now switching to OpenAPI, and not having "native" Kotlin support was one of the reasons for that (though not the only one).

As to the battle-testedness: I think if you compare Thrifty to some of the official language implementations from Apache, Thrifty would win in terms of users. And server support is really just some glue (result wrappers - I think the relevant portion of my changes amount to some 300 lines, not more) on top of what Thrifty already provides, so I am pretty confident in it.
My branch passes the "official" Apache Java conformance test (it's just the existing conformance test in reverse) apart from a small issue.

So I am closing this for now and will get back to you with the server support PR when I have time to continue.

**Comment by benjamin-bader** [comment]
> As to the battle-testedness: I think if you compare Thrifty to some of the official language implementations from Apache, Thrifty would win in terms of users.

I agree with you that Thrifty is more battle-tested than some of the Apache Thrift implementations, but I don't think that holds for the Java one.  It's in heavy and widespread use throughout industry, and at Facebook in particular.  

> And server support is really just some glue (result wrappers - I think the relevant portion of my changes amount to some 300 lines, not more) on top of what Thrifty already provides, so I am pretty confident in it.
My branch passes the "official" Apache Java conformance test (it's basically the existing conformance test in reverse) apart from a small issue.

That's really encouraging to hear about the conformance tests!  I think however you may be underestimating just how much work has gone in to optimizing the server components of Apache Thrift.  There are many non-obvious decisions, and certainly more than 300 lines of code to properly support e.g. async calls with proper timeout behavior in a performant fashion.

> (unless you can say right now that I am wasting my time).

You certainly aren't!  I don't mean to be discouraging, but rather to ask questions that might (or might not) shed light.  It's wonderful to see someone step up and propose new and interesting directions for the project.  I'd never considered doing a server implementation because a) we never needed it, and b) the problems Thrifty set out to solve were exclusive to Android (as you've noted).  You correctly observe that we've moved beyond that initial method-count problem, both as a project as as in industry (multidex is way less painful that it once was).  Certainly Kotlin offers some real advantages to data modelling.

I'm interested to see where the server support ends up; please continue.

**Comment by luqasn** [comment]
You weren't discouraging, but sometimes there are just features you don't want because they are a burden to maintain or don't fit well. So it's good to talk about it.

My current plan is not to support a full thrift Server out of the box, merely provide a _Processor_ implementation for a service that takes input and output protocol as parameters and delegates handling of the methods to a user-provided implementation of the interface.

So users would need to wrap their own http/socket handler around that processor. That is exactly the way we already use the "original" thrift server because you can't plug the provided servlet into ktor.

**Comment by benjamin-bader** [comment]
Incidentally, Apache's `TProcessor` interface sounds nearly exactly like what you're describing.  As far as I'm aware, sending thrift over plain old TCP sockets is the usual approach; `TServer` and its descendants all implement variations on the "open a socket, feed requests to a Processor` notion.  So, in comparison to Apache, it sounds like you want to focus more on the integration between Thrift services with servers like ktor (or jetty, netty, etc) than with actually implementing an RPC server.  That's a fine place to begin IMO; the `Processor` abstraction will allow for expanding the scope of the server work as needed.
